### ISSUE TITLE
Public API for Customizing Kernel Restart Behavior

### ISSUE THREAD

**Issue by ssanderson** [jupyter/jupyter_client#186]
On Quantopian, we regularly (multiple times per day at our current usage) run into the following pathological case with the default kernel restart behavior:
1. A user (running in a docker container launched by Dockerspawner) has many notebooks open, each of which consumes a significant amount of memory.
2. The user opens a new notebook, and the creation of their kernel hits a memory limit configured in docker, causing it to get OOM-killed.  We load a nontrivial amount of data in some of our extension setup, so a user with tens of megabytes left under the cap can OOM before they get a chance to run another cell.  Depending on how happy AWS is on any given day, and on how nice docker is feeling, it can take 10+ seconds to for the kernel to die.
3. The KernelRestarter notices that the kernel process has died and restarts it.  The restart attempt counter is set to 1.
4. Kernel survives long enough that the next poll succeeds, and we reset the restart attempt counter to zero.
5. Kernel gets OOM-killed again.
6. Repeat steps 3-5 until someone manually intervenes.

The simplest solution for us to to increase the poll interval on KernelRestarter to be long enough that a nearly-OOM kernel reliably dies before we poll again.  This is tricky to tune, however, and it has negative UX impact if the kernel crashes for other reasons.  

What I'd really like to be able to do is to customize the kernel restart logic to include heuristics for detecting and giving up early in the nearly-OOM case.  As far as I can tell, however, there's no way to provide a custom subclass of `KernelRestarter` in a config file, because the initialization logic for `IOLoopKernelManager` looks like:

```
    _restarter = Instance('jupyter_client.ioloop.IOLoopKernelRestarter', allow_none=True)

    def start_restarter(self):
        if self.autorestart and self.has_kernel:
            if self._restarter is None:
                self._restarter = IOLoopKernelRestarter(
                    kernel_manager=self, loop=self.loop,
                    parent=self, log=self.log
                )
            self._restarter.start()

    def stop_restarter(self):
        if self.autorestart:
            if self._restarter is not None:
                self._restarter.stop()
```

So I think with the way things stand right now, my best option is to override `MultiKernelManager.kernel_manager_class`, which is explicitly configurable.  Two questions about this:
1. Is it reasonable to override the whole kernel manager to customize the restart logic?
2. Is there interest in a public API for just configuring the restarting behavior?

**Comment by Carreau** [comment]
> Is it reasonable to override the whole kernel manager to customize the restart logic?

Thats seem like extra-complicated, if we can make your life simpler....

> Is there interest in a public API for just configuring the restarting behavior?

@yuvipanda , @minrk ; I guess in the context of the hub that might be often desired ?

**Comment by ssanderson** [comment]
I think the most straightforward fix would just be to make the restarter class on `KernelManager` configurable.

I ended up making this work by implementing my own subclass of `IOLoopKernelRestarter` and working around the non-configurability of the restarter by implementing a thin subclass of `IOLoopKernelManager` as well.  The whole setup currently looks like this:

```python

class ThrottledIOLoopKernelRestarter(IOLoopKernelRestarter):
    """
    Monitors and autorestarts a kernel using a throttling mechanism to limit
    restarts instead of checking for consecutive restarts as the default
    implementation does.

    A kernel starts with a budget of ``initial_restart_budget``.

    When the kernel is auto-restarted, we decrement ``restart_budget`` by 1.

    Every ``restart_budget_replenish_interval`` seconds, we increment
    ``restart_budget`` by 1 (capped at the original restart_budget).

    If ``restart_budget`` ever hits zero, we consider the kernel unrecoverable
    and stop attempting to restart it.
    """
    initial_restart_budget = Integer(
        default_value=3,
        allow_none=False,
        config=True
    )
    restart_budget = Integer(allow_none=False, config=False)

    def _restart_budget_default(self):
        return self.initial_restart_budget

    restart_budget_replenish_interval = Float(
        default_value=300.0,
        allow_none=False,
        config=True,
    )

    kernel_id = Unicode('', config=False)

    def _kernel_id_default(self):
        connection_file = self.kernel_manager.connection_file
        # Connection files look like '/path/to/kernel-<id>.json'.
        return os.path.basename(connection_file).rsplit('.', 1)[0]

    _replenish_callback = None

    def start(self):
        """Called on startup."""
        # Parent class starts a periodic callback of self.poll().
        super(ThrottledIOLoopKernelRestarter, self).start()

        # XXX: This is (I think?) defending against repeated calls of start()?
        # It's not clear whether this is an expected case, but the parent class
        # in IPython/Jupyter does this, so we do it too.
        if self._replenish_callback is None:
            self._replenish_callback = ioloop.PeriodicCallback(
                self.replenish_restart_budget,
                1000 * self.restart_budget_replenish_interval,
                io_loop=self.loop,  # loop inherited from parent.
            )
            self._replenish_callback.start()

    def stop(self):
        """Called on shutdown, or when we give up trying to restart."""
        # XXX: See note in start() re: guarding against repeated calls.
        if self._replenish_callback is not None:
            self._replenish_callback.stop()
            self._replenish_callback = None

        # Parent class stops the periodic callback of self.poll().
        super(ThrottledIOLoopKernelRestarter, self).stop()

    def poll(self):
        """Check if the kernel is alive.

        If it's dead and we have restart attempts left, attempt a restart and
        decrement our attempts count.

        If it's dead and we're out of restart attempts, give up.
        """
        if self.kernel_manager.is_alive():
            # Happy case: kernel is alive.
            return

        if self.restart_budget > 0:
            self.attempt_restart()
        else:
            self.give_up()

    def attempt_restart(self):
        """Attempt to restart the kernel."""
        self.log.info(
            '%s: restarting %s (%d attempt(s) left)',
            type(self).__name__,
            self.kernel_id,
            self.restart_budget,
        )
        self.restart_budget -= 1

        # now=True tells the manager not to wait for the kernel to exit
        # cleanly, since we know the kernel is already dead if we're here.
        self.kernel_manager.restart_kernel(now=True)

    def give_up(self):
        """Stop attempting to restart the kernel."""
        self.log.warn(
            "%s: giving up on reviving %s.",
            type(self).__name__,
            self.kernel_id,
        )
        self._fire_callbacks('dead')
        self.stop()

    def replenish_restart_budget(self):
        """Replenish our restart budget.

        This is scheduled as a periodic callback by `start()`.
        """
        if self.restart_budget < self.initial_restart_budget:
            self.restart_budget += 1
            self.log.info(
                "%s: replenished restart budget to %d",
                type(self).__name__,
                self.restart_budget,
            )


class ThrottledRestartsIOLoopKernelManager(IOLoopKernelManager):
    """An IOLoopKernelManager that uses ThrottledIOLoopKernelRestarter
    """
    # These two methods are almost entirely copied from IOLoopKernelManager.
    # The only difference is that we use ThrottledIOLoopKernelRestarter instead
    # of the IOLoopKernelRestarter.
    def start_restarter(self):
        if self.autorestart and self.has_kernel:
            if self._restarter is None:
                self._restarter = ThrottledIOLoopKernelRestarter(
                    kernel_manager=self,
                    loop=self.loop,
                    parent=self,
                    log=self.log
                )
            self._restarter.start()

    def stop_restarter(self):
        if self.autorestart:
            if self._restarter is not None:
                self._restarter.stop()
```

**Comment by minrk** [comment]
I think the whole KernelManager API needs a pass with an eye for subclassing. These are being increasingly common, and the API wasn't written with that in mind. Your help with that would be super appreciated. Making the restarter_class configurable is certainly a simple place to start.

**Comment by ssanderson** [comment]
@minrk I missed your last comment here.  I'll try to take a pass at making the restarter configurable this weekend.

**Comment by ssanderson** [comment]
...aaaand over a year later: https://github.com/jupyter/jupyter_client/pull/290/files

**Comment by takluyver** [comment]
Closed by #290 (as I understand it - let me know if there's more to be done on this)
